tempesta_fw/http_parser.c

```diff
@@ -2089,16 +2090,20 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
 	TfwHttpReq *req = (TfwHttpReq *)req_data;
 	__FSM_DECLARE_VARS(req);
 
-	__hbh_parser_init_req(req);
-
 	TFW_DBG("parse %lu client data bytes (%.*s%s) on req=%p\n",
 		len, min(500, (int)len), data, len > 500 ? "..." : "", req);
 
 	__FSM_START(parser->state) {
 
 	/* ----------------    Request Line    ---------------- */
 
+	/* Parser internal initilizers, must be called once per message. */
 	__FSM_STATE(Req_0) {
+		__hbh_parser_init_req(req);
+		/* fall through */
```

Comment:
I have checked out all the places in code, where _fall through_ used instead of `FSM_MOVE` directives. No errors there and  _fall through_ can be replaced with `__FSM_JMP`. Since desired state is the next one _fall through_ is used.

`__FSM_STATE` saves current parser state to `__fsm_const_state`: https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/http_parser.c#L120. In all the places in parser _fall through_ is used only when we do not move our current position in message and we want to work with the same data in next stage or stages. 

Any flavour of `FSM_MOVE*` is not an option here since `parser->to_go` will be set to 0 for situation you are describing ( https://github.com/tempesta-tech/tempesta/blob/master/tempesta_fw/http_parser.c#L137 ). And next stage will take *next* character, not the same one.